### ISSUE TITLE
Array of event handlers in config

### ISSUE THREAD

**Issue by paulzi** [yiisoft/yii2#11269]
In `common\config\main.php` i'm use this config for :

``` php
   'components' => [ /* ... */ ],
   'on beforeRequest' => function () {
        \Yii::$container->set('yii\widgets\LinkPager', ['maxButtonCount' => 5]);
    },
```

But in `frontend\config\main.php` i'm need use second handler for styling Html:

``` php
   'components' => [ /* ... */ ],
   'on beforeRequest' => function () {
        Yii::$classMap['yii\helpers\Html'] = '@common/helpers/Html.php';
    },
```

Common config does not apply, because it uses `ArrayHelper::merge()` and `on beforeRequest` is overwrite.
## Suggestion for improvement

It would be nice if as values could be to use an array:

``` php
   'components' => [ /* ... */ ],
   'on beforeRequest' => [
      function () {
         \Yii::$container->set('yii\widgets\LinkPager', ['maxButtonCount' => 5]);
      }
   ],
```

| Q | A |
| --- | --- |
| Yii version | 2.0.7 |

**Comment by mdmunir** [comment]
create your own behavior
```php
class MyBehavior extends Behavior
{
    public $events = [];
    public function attach($owner)
    {
        $this->owner = $owner;
        foreach ($this->events as $handler) {
            $owner->on('beforeRequest', $handler);
        }
    }
}
```

`common/config/main.php`
```php
'as xxx' => [
    'class' => 'common\classes\MyBehavior',
    'events' => [function(){ ... }],
]
```

`frontend/config/main.php`
```php
'as xxx' => [
    'class' => 'common\classes\MyBehavior',
    'events' => [function(){ ... }],
]
```

**Comment by paulzi** [comment]
Thank you for this decision, but it is ugly. I need to create behavior, and use two different names to connect `as xxx` and `as yyy`.

**Comment by dynasource** [comment]
> Common config does not apply, because it uses ArrayHelper::merge() and on beforeRequest is overwrite.

This is valid usecase. Its hard to merge.

**Comment by SilverFire** [comment]
Separate behaviors that provide handlers for specific events contains more semantic, so it's more preferred.
Considering problem is not popular and in PR #13199 we have not find a reliable way to do it, I'm closing this issue. Thank you

**Comment by rob006** [comment]
@SilverFire 

> Separate behaviors that provide handlers for specific events contains more semantic, so it's more preferred.

Behaviors has much more overhead and there are also plans to remove them (or at least prefer events over behaviors where possible): https://github.com/yiisoft/yii2/wiki/Plan-for-next-major-versions#core

> Considering problem is not popular and in PR #13199 we have not find a reliable way to do it

I've already proposed [solution](https://github.com/yiisoft/yii2/pull/13199#issuecomment-267001421) (checking if value is array of closures) - what is unreliable in it?

**Comment by cebe** [comment]
> I've already proposed solution (checking if value is array of closures) - what is unreliable in it?

that sounds like a possible solution, had not seen that comment.

**Comment by SilverFire** [comment]
> I've already proposed solution (checking if value is array of closures) - what is unreliable in it?

That comment is about `at least`, that usually brings inconsistency.

**Comment by SilverFire** [comment]
Hm... We can use `Closure::fromCallable()`

**Comment by rob006** [comment]
> That comment is about at least, that usually brings inconsistency.

There are many places where closures are supported and other callable types are not. The only way to fix this is force `Closure` everywhere.

**Comment by samdark** [comment]
@SilverFire fromCallable is PHP 7.1+ so 2.1 only.

**Comment by samdark** [comment]
Won't be implemented in 2.0.
